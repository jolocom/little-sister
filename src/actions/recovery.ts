import { ThunkAction } from '../store'
import { SoftwareKeyProvider } from 'jolocom-lib/js/vaultedKeyProvider/softwareProvider'
import { navigationActions } from './index'
import { routeList } from '../routeList'
import settingKeys from '../ui/settings/settingKeys'
import { SETTINGS } from '../reducers/settings'
import Share from 'react-native-share'
import { toBase64 } from '../lib/util'
import { removeNotification } from './notifications'

export const showSeedPhrase = (): ThunkAction => async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  const encryptedSeed = await backendMiddleware.storageLib.get.encryptedSeed()
  if (!encryptedSeed) {
    throw new Error('Can not retrieve Seed from database')
  }
  // TODO create vault from encrypted Seed
  const pass = await backendMiddleware.keyChainLib.getPassword()
  const vault = new SoftwareKeyProvider(Buffer.from(encryptedSeed, 'hex'))
  const mnemonic = vault.getMnemonic(pass)
  return dispatch(
    navigationActions.navigate({
      routeName: routeList.SeedPhrase,
      params: { mnemonic },
    }),
  )
}

export const setSeedPhraseSaved = (): ThunkAction => async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  await backendMiddleware.storageLib.store.setting(
    settingKeys.seedPhraseSaved,
    true,
  )

  // TODO: find sticky by id from queue, not active
  const {
    notifications: { active: stickyNotification },
  } = getState()
  if (stickyNotification) dispatch(removeNotification(stickyNotification))

  return dispatch({
    type: SETTINGS.SET_SEED_PHRASE_SAVED,
  })
}

export const setAutoBackup = (
  isEnabled: boolean,
  shouldUpload: boolean = true,
): ThunkAction => async (dispatch, getState, backendMiddleware) => {
  await backendMiddleware.storageLib.store.setting(
    settingKeys.autoBackup,
    isEnabled,
  )
  await dispatch({
    type: SETTINGS.SET_AUTO_BACKUP,
    value: isEnabled,
  })
  // initial backup creation and upload (after store is updated)
  if (isEnabled && shouldUpload) await dispatch(autoBackupData())
}

// if no value is provided the current date-time is used
export const setLastBackup = (value?: string): ThunkAction => async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  const currDate = value === undefined ? new Date().toISOString() : value
  await backendMiddleware.storageLib.store.setting(
    settingKeys.lastBackup,
    currDate,
  )
  return dispatch({
    type: SETTINGS.SET_LAST_BACKUP,
    value: currDate,
  })
}

export const disableAndRemoveBackup = (): ThunkAction => async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  dispatch(setAutoBackup(false, false))
  dispatch(setLastBackup(''))
  await backendMiddleware.deleteBackup()
}

export const manualBackup = (): ThunkAction => async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  const encryptedBackup = await backendMiddleware.backupData(false)

  // throws if user dismissed the share dialog (sharing is assumed to be successful
  // as soon as an app is selected --> not always "really" successful)
  await Share.open({
    filename: 'jolocom-backup', // not working currently, will work with react-native-share >=2.0.0 https://github.com/react-native-community/react-native-share/pull/565
    url: `data:text/plain;base64,${toBase64(JSON.stringify(encryptedBackup))}`,
  })
  // in case something throws (e.g. upload fails, user dismissed) this code will not be reached
  dispatch(setLastBackup())
}

export const autoBackupData = (): ThunkAction => async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  if (getState().settings.autoBackup) await backendMiddleware.backupData(true)
  dispatch(setLastBackup())
}
