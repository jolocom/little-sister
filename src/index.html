<!DOCTYPE html>

<html>
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport"
          content="width=device-width, initial-scale=1, user-scalable=no">

    <title>Little Sister</title>

    <link href="https://fonts.googleapis.com/css?family=Roboto:100,400,300,600"
          rel="stylesheet"
          type="text/css" />
    <link href="https://fonts.googleapis.com/icon?family=Material+Icons"
          rel="stylesheet">

    <style type="text/css">
      * {
        margin: 0;
        padding: 0;
      }
      html, body {
        height: 100%;
      }
      body {
        font-family: Roboto;
      }
      #app {
        width: 100%;
        height: 100%;
      }
    </style>

    <!-- <meta http-equiv="Content-Security-Policy" content="default-src * data: gap: https://ssl.gstatic.com 'unsafe-eval'; style-src * 'unsafe-inline'; media-src *; img-src * data:"> -->
    <meta name="format-detection" content="telephone=no">
    <meta name="msapplication-tap-highlight" content="no">
    <meta name="viewport" content="user-scalable=no, initial-scale=1, maximum-scale=1, minimum-scale=1, width=device-width">
  </head>
  <body>
    <div id="app"></div>
    <script src="https://solid.github.io/releases/rdflib.js/rdflib-0.7.0.min.js"></script>
    <script src="https://d3js.org/d3.v4.min.js"></script>
    <script src="js/bundle.js"></script>
  </body>
</html>
