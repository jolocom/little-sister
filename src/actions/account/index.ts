import { AnyAction, Dispatch } from 'redux'
import { genericActions, navigationActions } from 'src/actions/'
import { BackendMiddleware } from 'src/backendMiddleware'
import { routeList } from 'src/routeList'
import { DecoratedClaims, CategorizedClaims } from 'src/reducers/account'
import { SignedCredential } from 'jolocom-lib/js/credentials/signedCredential/signedCredential'
import {
  getClaimMetadataByCredentialType,
  getCredentialUiCategory,
  getUiCredentialTypeByType,
  instantiateIdentityWallet,
} from '../../lib/util'
import { cancelReceiving } from '../sso'

export const setDid = (did: string) => ({
  type: 'DID_SET',
  value: did,
})

export const setSelected = (claim: DecoratedClaims) => ({
  type: 'SET_SELECTED',
  selected: claim,
})

export const resetSelected = () => ({
  type: 'RESET_SELECTED',
})

export const handleClaimInput = (fieldValue: string, fieldName: string) => ({
  type: 'HANLDE_CLAIM_INPUT',
  fieldName,
  fieldValue,
})

export const toggleClaimsLoading = (value: boolean) => ({
  type: 'TOGGLE_CLAIMS_LOADING',
  value,
})

export const checkIdentityExists = () => async (
  dispatch: Dispatch<AnyAction>,
  getState: Function,
  backendMiddleware: BackendMiddleware,
) => {
  const { storageLib } = backendMiddleware

  try {
    const personas = await storageLib.get.persona()
    if (!personas.length) {
      dispatch(toggleLoading(false))
      return
    }

    dispatch(setDid(personas[0].did))
    await instantiateIdentityWallet(backendMiddleware)

    dispatch(toggleLoading(false))
    dispatch(navigationActions.navigatorReset({ routeName: routeList.Home }))
  } catch (err) {
    if (err.message.indexOf('no such table') === 0) {
      return
    }
    dispatch(genericActions.showErrorScreen(err))
  }
}

export const setIdentityWallet = () => async (
  dispatch: Dispatch<AnyAction>,
  getState: Function,
  backendMiddleware: BackendMiddleware,
) => {
  try {
    await instantiateIdentityWallet(backendMiddleware)
  } catch (err) {
    dispatch(genericActions.showErrorScreen(err))
  }
}

export const openClaimDetails = (claim: DecoratedClaims) => (
  dispatch: Dispatch<AnyAction>,
) => {
  dispatch(setSelected(claim))
  dispatch(
    navigationActions.navigate({
      routeName: routeList.ClaimDetails,
    }),
  )
}

export const saveClaim = () => async (
  dispatch: Dispatch<AnyAction>,
  getState: Function,
  backendMiddleware: BackendMiddleware,
) => {
  const { identityWallet, storageLib, keyChainLib } = backendMiddleware

  try {
    const did = getState().account.did.get('did')
    const claimsItem = getState().account.claims.toJS().selected
    const password = await keyChainLib.getPassword()

    const verifiableCredential = await identityWallet.create.signedCredential(
      {
        metadata: getClaimMetadataByCredentialType(claimsItem.credentialType),
        claim: claimsItem.claimData,
        subject: did,
      },
      password,
    )

    if (claimsItem.id) {
      await storageLib.delete.verifiableCredential(claimsItem.id)
    }

    await storageLib.store.verifiableCredential(verifiableCredential)
    await setClaimsForDid()

    dispatch(
      navigationActions.navigatorReset({
        routeName: routeList.Home,
      }),
    )
  } catch (err) {
    dispatch(genericActions.showErrorScreen(err))
  }
}

// TODO Currently only rendering  / adding one
export const saveExternalCredentials = () => async (
  dispatch: Dispatch<AnyAction>,
  getState: Function,
  backendMiddleware: BackendMiddleware,
) => {
  const { storageLib } = backendMiddleware
  const externalCredentials = getState().account.claims.toJS().pendingExternal
  const cred: SignedCredential = externalCredentials[0]

  if (cred.id) {
    await storageLib.delete.verifiableCredential(cred.id)
  }

  try {
    await storageLib.store.verifiableCredential(externalCredentials[0])
    dispatch(cancelReceiving())
  } catch (err) {
    dispatch(genericActions.showErrorScreen(err))
  }
}

export const toggleLoading = (value: boolean) => ({
  type: 'SET_LOADING',
  value,
})

export const setClaimsForDid = () => async (
  dispatch: Dispatch<AnyAction>,
  getState: Function,
  backendMiddleware: BackendMiddleware,
) => {
  dispatch(toggleClaimsLoading(true))
  const storageLib = backendMiddleware.storageLib

  const verifiableCredentials: SignedCredential[] = await storageLib.get.verifiableCredential()
  const claims = prepareClaimsForState(
    verifiableCredentials,
  ) as CategorizedClaims

  dispatch({
    type: 'SET_CLAIMS_FOR_DID',
    claims,
  })

  dispatch(toggleClaimsLoading(false))
}

const prepareClaimsForState = (credentials: SignedCredential[]) => {
  const categorizedClaims = {}
  const decoratedCredentials = convertToDecoratedClaim(credentials)

  decoratedCredentials.forEach(decoratedCred => {
    const uiCategory = getCredentialUiCategory(decoratedCred.credentialType)

    try {
      categorizedClaims[uiCategory].push(decoratedCred)
    } catch (err) {
      categorizedClaims[uiCategory] = [decoratedCred]
    }
  })

  return categorizedClaims
}

// TODO Util, make subject mandatory
export const convertToDecoratedClaim = (
  vCreds: SignedCredential[],
): DecoratedClaims[] =>
  vCreds.map(vCred => {
    const claimData = { ...vCred.claim }
    delete claimData.id

    return {
      credentialType: getUiCredentialTypeByType(vCred.type),
      claimData,
      id: vCred.id,
      issuer: vCred.issuer,
      subject: vCred.claim.id || 'Not found',
      expires: vCred.expires || undefined,
    }
  })
