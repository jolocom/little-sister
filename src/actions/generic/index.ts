import { navigationActions } from 'src/actions/'
import { routeList } from 'src/routeList'
import SplashScreen from 'react-native-splash-screen'
import I18n from 'src/locales/i18n'
import { ThunkAction } from 'src/store'
import { AppError, ErrorCode } from 'src/lib/errors'
import settingKeys from '../../ui/settings/settingKeys'
import { SETTINGS } from '../../reducers/settings'

export const showErrorScreen = (
  error: AppError | Error,
): ThunkAction => dispatch => {
  const appError: AppError =
    error.constructor === AppError
      ? (error as AppError)
      : new AppError(ErrorCode.Unknown, error)

  return dispatch(
    navigationActions.navigate({
      routeName: routeList.Exception,
      params: {
        returnTo: appError.navigateTo,
        error: appError,
      },
    }),
  )
}

export const initApp: ThunkAction = async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  try {
    await backendMiddleware.initStorage()
    const storedSettings = await backendMiddleware.storageLib.get.settingsObject()
    // locale setup
    if (storedSettings.locale) I18n.locale = storedSettings.locale
    else storedSettings.locale = I18n.locale
    SplashScreen.hide()
    return dispatch(loadSettings(storedSettings))
  } catch (e) {
    return dispatch(
      showErrorScreen(
        new AppError(ErrorCode.WalletInitFailed, e, routeList.Landing),
      ),
    )
  }
}

export const loadSettings = (settings: { [key: string]: any }) => ({
  type: SETTINGS.LOAD_SETTINGS,
  value: settings,
})

export const setLocale = (locale: string): ThunkAction => async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  await backendMiddleware.storageLib.store.setting(settingKeys.locale, locale)
  I18n.locale = locale
  dispatch({ type: SETTINGS.SET_LOCALE, value: locale })

  // we need to reset the navigator so that all screens are re-rendered with the
  // new locale
  return dispatch(navigationActions.navigatorReset())
}
