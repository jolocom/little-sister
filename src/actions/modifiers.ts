import { ActionCreator, AnyAction } from 'redux'
import NetInfo from '@react-native-community/netinfo'
import { ThunkAction } from 'src/store'
import { AppError, ErrorCode } from '../lib/errors'
import { toggleLoading } from './account'
import { showErrorScreen } from './generic'
import {
  scheduleErrorNotification,
  scheduleOfflineNotification,
} from './notifications/creators'

/**
 * Curried function that wraps a {@link ThunkAction} with two calls to the provided loadingAction
 * {@link ActionCreator} (called with "true" before the wrapped action is dispatched, and with "false" after)
 * @param loadingAction - An action creator to create the actions to be dispatched before and after the wrappedAction-
 * @param wrappedAction - The thunkAction to be wrapped
 * @example dispatch(withLoading(toggleLoading)(saveClaims))
 */
export const withLoadingHandler = (loadingAction: ActionCreator<AnyAction>) => (
  wrappedAction: ThunkAction,
): ThunkAction => async dispatch => {
  try {
    dispatch(loadingAction(true))
    return await dispatch(wrappedAction)
  } finally {
    // NOTE: timeout is a hack to avoid flashing when navigating, as the new
    // screen will have not been loaded yet but the AppLoading will be taken
    // down, which will flash the old screen shortly
    setTimeout(() => dispatch(loadingAction(false)), 100)
  }
}

/**
 * Curried function that wraps a {@link ThunkAction} with a notification on internet connection absence
 * @param notificationAction - The action that should schedule a notification when there is no connection present
 * @param wrappedAction - The thunkAction to be wrapped
 * @example dispatch(withInternet((saveClaims))
 */
export const withInternetHandler = (notificationAction: ThunkAction) => (
  wrappedAction: ThunkAction,
): ThunkAction => async dispatch => {
  const state = await NetInfo.fetch()
  if (!state.isConnected) {
    return dispatch(notificationAction)
  }
  return dispatch(wrappedAction)
}

type ErrorModifier = (error: AppError | Error) => AppError

/**
 * Curried function that wraps a {@link ThunkAction} with the provided error handler to be dispatched on thrown error
 * @param errorHandler - An {@link ActionCreator} to be dispatched with the error if an error is thrown
 * @param errorModifier - An optional modifier for the error which is thrown, is Identity function by default
 * @param wrappedAction - The {@link ThunkAction} to be wrapped. If dispatching it throws an error, the errorhandler will be dispatched
 * @example dispatch(withErrorHandling(showErrorScreen)(saveClaims))
 * @dev The return value from this modifier can be passed to other modifiers, i.e. modifiers can be composed
 * e.g. dispatch(withLoading(toggleLoading)(withErrorHandling(showErrorScreen)(saveClaims)))
 */
export const withErrorHandler = (
  errorHandler: ActionCreator<ThunkAction>,
  errorModifier: ErrorModifier | undefined = undefined,
) => (
  wrappedAction: ThunkAction,
  modifier: ErrorModifier | undefined = undefined,
): ThunkAction => async dispatch => {
  modifier = modifier || errorModifier
  try {
    return await dispatch(wrappedAction)
  } catch (error) {
    if (modifier) error = modifier(error)
    return dispatch(errorHandler(error))
  }
}

const showErrorNotification = (
  error: AppError | Error,
): ThunkAction => dispatch => {
  const appError: AppError =
    error.constructor === AppError
      ? (error as AppError)
      : new AppError(ErrorCode.Unknown, error)

  return dispatch(scheduleErrorNotification(appError))
}

export const withLoading = withLoadingHandler(toggleLoading)
export const withErrorScreen = withErrorHandler(showErrorScreen)
export const withInternet = withInternetHandler(scheduleOfflineNotification)
export const withErrorNotification = withErrorHandler(showErrorNotification)
