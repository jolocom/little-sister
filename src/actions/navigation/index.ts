import {
  NavigationActions,
  NavigationNavigateActionPayload,
} from 'react-navigation'
import { AnyAction, Dispatch } from 'redux'
import { ssoActions } from 'src/actions/'
import { setDid, toggleLoading } from '../account'
import { BackendMiddleware } from 'src/backendMiddleware'
import { instantiateIdentityWallet } from 'src/lib/util'
import { setDeepLinkLoading } from '../sso'

export const navigate = (options: NavigationNavigateActionPayload) =>
  NavigationActions.navigate(options)

export const goBack = () => NavigationActions.back()

export const navigatorReset = (newScreen: NavigationNavigateActionPayload) =>
  NavigationActions.reset({
    index: 0,
    actions: [navigate(newScreen)],
  })

/**
 * The function that parses a deep link to get the route name and params
 * It then matches the route name and dispatches a correcponding action
 * @param url - a deep link string with the following schemat: appName://routeName/params
 */
export const handleDeepLink = (url: string) => async (
  dispatch: Dispatch<AnyAction>,
  getState: Function,
  backendMiddleware: BackendMiddleware,
) => {
  dispatch(toggleLoading(true))
  const route: string = url.replace(/.*?:\/\//g, '')
  const params: string = (route.match(/\/([^\/]+)\/?$/) as string[])[1] || ''
  const routeName = route!.split('/')[0]

  if (
    routeName === 'consent' ||
    routeName === 'payment' ||
    routeName === 'authenticate'
  ) {
    dispatch(setDeepLinkLoading(true))
    const personas = await backendMiddleware.storageLib.get.persona()

    if (!personas.length) {
      dispatch(toggleLoading(false))
      return
    }

    dispatch(setDid(personas[0].did))
    await instantiateIdentityWallet(backendMiddleware)
    dispatch(ssoActions.parseJWT(params))
  }
}
