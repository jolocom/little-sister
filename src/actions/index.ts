import * as accountActions from 'src/actions/account/'
import * as registrationActions from 'src/actions/registration/'
import * as navigationActions from 'src/actions/navigation/'
import * as genericActions from 'src/actions/generic/'
import * as ssoActions from 'src/actions/sso/'

export {
  accountActions,
  registrationActions,
  navigationActions,
  genericActions,
  ssoActions,
}
