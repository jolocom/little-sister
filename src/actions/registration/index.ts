import { navigationActions } from 'src/actions/'
import { routeList } from 'src/routeList'
import * as loading from 'src/actions/registration/loadingStages'
import { setDid } from 'src/actions/account'
import { ThunkAction } from 'src/store'
import { navigatorResetHome } from '../navigation'
import { setAutoBackup, setSeedPhraseSaved } from '../recovery'
import { EncryptedData } from 'jolocom-lib/js/vaultedKeyProvider/softwareProvider'

export const setLoadingMsg = (loadingMsg: string) => ({
  type: 'SET_LOADING_MSG',
  value: loadingMsg,
})

export const setIsRegistering = (value: boolean) => ({
  type: 'SET_IS_REGISTERING',
  value,
})

export const createIdentity = (encodedEntropy: string): ThunkAction => async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  dispatch(
    navigationActions.navigate({
      routeName: routeList.RegistrationProgress,
    }),
  )

  const isRegistering = getState().registration.loading.isRegistering
  if (isRegistering) return

  dispatch(setIsRegistering(true))

  dispatch(setLoadingMsg(loading.loadingStages[0]))
  await backendMiddleware.createKeyProvider(encodedEntropy)

  dispatch(setLoadingMsg(loading.loadingStages[1]))
  await backendMiddleware.fuelKeyWithEther()

  dispatch(setLoadingMsg(loading.loadingStages[2]))
  const identity = await backendMiddleware.createIdentity()

  dispatch(setDid(identity.did))
  dispatch(setLoadingMsg(loading.loadingStages[3]))
  dispatch(setIsRegistering(false))

  return dispatch(navigatorResetHome())
}

export const recoverFromSeedPhrase = (mnemonic: string): ThunkAction => async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  dispatch(setIsRegistering(true))
  let pubKey
  try {
    pubKey = await backendMiddleware.recoverKeyProvider(mnemonic)
  } catch (e) {
    return dispatch(setIsRegistering(false))
  }

  const backup = await backendMiddleware.fetchBackup()
  if (backup) {
    dispatch(setAutoBackup(true, false))
    return dispatch(recoverIdentity(backup))
  }

  // No backup found, user needs to import manually
  // public key is needed to validate if the imported backup belongs to the recovered identity
  return dispatch(
    navigationActions.navigate({
      routeName: routeList.ImportBackup,
      params: { pubKey },
    }),
  )
}

export const recoverIdentity = (
  encryptedBackup?: EncryptedData,
): ThunkAction => async (dispatch, getState, backendMiddleware) => {
  // the backup should be validated before to avoid decryption failures at this point
  const backup = encryptedBackup
    ? await backendMiddleware.decryptBackup(encryptedBackup)
    : undefined
  const identity = await backendMiddleware.recoverIdentity(
    backup ? backup.did : undefined,
  )
  if (backup) await backendMiddleware.recoverData(backup)

  await dispatch(setDid(identity.did))
  await dispatch(setSeedPhraseSaved())
  await dispatch(setIsRegistering(false))
  return dispatch(navigatorResetHome())
}
