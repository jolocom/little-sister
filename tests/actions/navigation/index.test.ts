import { navigationActions, ssoActions } from '../../../src/actions'

import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'

describe('Navigation action creators', () => {
  describe('handleDeepLink', () => {
    const jwt =
      'eyJ0eXAiOiJKV1QiLCJhbGciOiJFUzI1NksifQ.eyJpbnRlcmFjdGlvblRva2VuIjp7'

    // TODO: refactor test case to account for identity check when deeplinking
    it('should extract the route name and param from the URL', async () => {
      const mockStore = configureStore([thunk])({})
      const mockBackendMiddleware = {
        storageLib: {
          get: {
            persona: jest.fn().mockResolvedValue([{ did: 'did:jolo:mock' }]),
            encryptedSeed: jest.fn().mockResolvedValue('johnnycryptoseed'),
          },
        },
        keyChainLib: {
          getPassword: jest.fn().mockResolvedValue('secret123'),
        },
        encryptionLib: {
          decryptWithPass: () => 'angelaMerkleTreeSeed',
        },
        setIdentityWallet: jest.fn(() => Promise.resolve()),
      }
      const parseJWTSpy = jest
        .spyOn(ssoActions, 'parseJWT')
        .mockImplementation(() => () => {})
      const action = navigationActions.handleDeepLink(
        'smartwallet://consent/' + jwt,
      )
      await action(mockStore.dispatch, () => {}, mockBackendMiddleware)
      expect(mockStore.getActions()).toMatchSnapshot()
      expect(parseJWTSpy).toHaveBeenCalledWith(jwt)
      parseJWTSpy.mockReset()
    })

    it('should not atttempt to parse if route was not correct', () => {
      const mockStore = configureStore([thunk])({})
      const parseJWTSpy = jest.spyOn(ssoActions, 'parseJWT')
      const action = navigationActions.handleDeepLink(
        'smartwallet://somethingElse/' + jwt,
      )

      action(mockStore.dispatch)
      expect(mockStore.getActions()).toMatchSnapshot()
      expect(parseJWTSpy).not.toHaveBeenCalled()
    })
  })
})
