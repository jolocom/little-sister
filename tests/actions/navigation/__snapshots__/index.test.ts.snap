// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation action creators handleDeepLink should extract the route name and param from the URL 1`] = `
Array [
  Object {
    "type": "SET_LOADING",
    "value": true,
  },
  Object {
    "type": "SET_DEEP_LINK_LOADING",
    "value": true,
  },
  Object {
    "type": "DID_SET",
    "value": "did:jolo:mock",
  },
]
`;

exports[`Navigation action creators handleDeepLink should not atttempt to parse if route was not correct 1`] = `
Array [
  Object {
    "type": "SET_LOADING",
    "value": true,
  },
]
`;
